### PR TITLE
fix: cache middleware - OKTA-280090

### PR DESCRIPTION
**Pull request by shuowu**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Okta Node SDK Changelog
 
+## 4.4.0
+
+- [218](https://github.com/okta/okta-sdk-nodejs/pull/218)
+  - Uses `req.url` as key to cache response
+  - Addes `json()` function to the default cache middleware response
```
Comment (swiftone) [inline CHANGELOG.md:7]: nit: Adds not Addes

File: src/http.js [hunk 2]
```diff
@@ -71,7 +71,7 @@ class Http {
 
     let retriedOnAuthError = false;
     const execute = () => {
-      const promise = this.prepareRequest(request)
+      const promise = () => this.prepareRequest(request)
```
Comment (swiftone) [inline src/http.js:74]: Can we change the variable name?  It's now a function that returns a promise, not a promise

File: test/unit/default-cache-middleware.js [hunk 3]
```diff
@@ -49,14 +46,16 @@ describe('Default cache middleware', () => {
...
       cacheStore
     };
-    const body = JSON.stringify(_.set({}, '_links.self.href', 'http://example.com/item'));
-    await cacheStore.set(ctx.req.uri, body);
+    const bodyObj = _.set({}, '_links.self.href', 'http://example.com/item');
+    const bodyStr = JSON.stringify(bodyObj);
+    await cacheStore.set(ctx.req.url, bodyStr);
     await middleware(ctx, () => next(ctx));
-    expect(await ctx.res.text()).to.equal(body);
+    expect(await ctx.res.json()).to.eql(bodyObj);
```
Comment (swiftone) [inline test/unit/default-cache-middleware.js:57]: eql()?   I couldn't find this matcher in the jest docs, what is it?

Comment (shuowu) [reply test/unit/default-cache-middleware.js:57]: tests under `test/unit/` are using chai, https://www.chaijs.com/api/bdd/#method_eql

File: src/http.js [hunk 4]
```diff
@@ -86,11 +86,11 @@ class Http {
         });
 
       if (!this.cacheMiddleware) {
-        return promise;
+        return promise();
       }
 
       const ctx = {
-        uri, // TODO: remove unused property. req.url should be the key.
+        uri, // TODO: remove unused property. req.url should be the key. OKTA-351525
```
Comment (aarongranick-okta) [inline src/http.js:93]: In this case, I think it is a case of a duplicate property. `ctx.uri` and `ctx.req.url` now hold the same value. The code in `default-cache-middleware` was looking for `ctx.req.uri` which does not exist and was always undefined. (typescript could have caught this, having a strongly typed context type)

We can choose which value should be used as the cache key. For some reason, which I don't recall, it was chosen to be `ctx.req.url` and now `ctx.uri` is duplicate/unused. We can keep it around for compatibility, or remove it.

File: src/default-cache-middleware.js [hunk 5]
```diff
@@ -15,15 +15,21 @@ const _ = require('lodash');
 module.exports = function defaultCacheMiddleware(ctx, next) {
   let cacheCheck, cacheHit = false;
   if (ctx.req.method.toLowerCase() === 'get' && !ctx.isCollection) {
-    // TODO: use ctx.req.url as cache key and add json() method to cached result: https://oktainc.atlassian.net/browse/OKTA-280090
-    cacheCheck = ctx.cacheStore.get(ctx.req.uri)
+    cacheCheck = ctx.cacheStore.get(ctx.req.url)
```
Comment (aarongranick-okta) [inline src/default-cache-middleware.js:18]: note: `ctx.req.uri` was a typo and so the value was always undefined. `ctx.req.url` holds a string value for the original uri of the request.

**Review by aarongranick-okta — APPROVED** [review]
It would be nice to have better testing. The same typo in the tests (`context.req.uri`) meant they passed. Would be great to have better testing for the cache feature. Maybe make two requests with the same URI and verify that only one mock XHR was called.

**Comment by shuowu** [comment]
> It would be nice to have better testing. The same typo in the tests (`context.req.uri`) meant they passed. Would be great to have better testing for the cache feature. Maybe make two requests with the same URI and verify that only one mock XHR was called.

Looks like you have handled it in test before, I just re-enabled it in this PR, https://github.com/okta/okta-sdk-nodejs/pull/218/files#diff-abdf0a4c5bf82c7aa5f922cead41f12e9e607b9c48bc3692b8b538865323e81dR356

File: src/http.js (hunk 4 above)
Comment (shuowu) [reply src/http.js:93]: It's kept for backward compatibility, as the users are able to provide other cache options that depend on the `ctx.uri` key. 

I created OKTA-351525 to remove it in the next major version update.

**Pull request closed without merging**